h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeNode.java

```diff
@@ -177,11 +177,15 @@ private BitSet calculateChildInclusiveLevels(boolean includeAllLevels,
     BitSet inheritedInclusiveLevels = findInclusiveLevels(colId);
     BitSet childInclusiveLevels = new BitSet();
 
+    int splitPoint = Float.isNaN(splitValue) ? -1 : (int) Math.ceil(splitValue);
+
     for (int i = 0; i < domainValues.length; i++) {
       // Calculate whether this level should flow into this child node.
       boolean includeThisLevel = false;
       {
-        if (discardAllLevels) {
+        if (splitPoint != -1) {
```

Comment:
Can we just work with the original `splitValue` here?

I think it would be a bit clearer if we just compared the index to the original splitValue, wdyt?

Reply:
It is correct that the `splitPoint != -1` check can be dropped, as any comparison of a Float compared to NaN yields false.

Reply:
Resolved in commit https://github.com/h2oai/h2o-3/pull/3534/commits/a82f14048dd9fb0078b555443d787b38522c513f

The commit message says:  `Simplified calculating child inclusive levels with numerically represented categorical split. The regimes discardAll and includeAll are now checked first.`

Thanks to you, I found a potential bug. There are two additional modes:

- include all categoricals
- discard all categoricals

Those in my opinion need to be checked first, as they could be potentially present with both numerical and bitset representation of the split. So I moved those a little bit "more up".

When all categoricals are not included nor discarded, a third check takes place: `else if (!Float.isNaN(splitValue)) ` - if the node has numerically represented categorical split, the inclusive categorical is calculated as `includeThisLevel = Math.ceil(splitValue) > i ^ nodeBitsetDoesContain;`.

If the split is not represented by a numerical split but rather by a `BitSet` (`else if (bs.isInRange(i) && bs.contains(i) == nodeBitsetDoesContain)`), original logic `includeThisLevel = calculateIncludeThisLevel(inheritedInclusiveLevels, i);` is used.

Reply:
I also got rid of the Math.ceil operation by inverting the problem from `Math.ceil(splitValue) > i ^ nodeBitsetDoesContain` to `splitValue < i ^ !nodeBitsetDoesContain;`

Should be faster, as the ceil operation is costly, the comparison operation remains and inverting a bit is cheaper.